cv_next/app/feed/components/filterPanel.tsx

```diff
@@ -16,14 +19,40 @@ export const FilterPanel = ({
   onChange: (filters: filterValues) => void;
 }) => {
   const [searchValue, setSearchValue] = useState(defaultFilters.searchValue);
-  const [categoryId, setCategoryId] = useState(defaultFilters.categoryIds);
+  const [categoryIds, setCategoryId] = useState(defaultFilters.categoryIds);
+
+  const searchParams = useSearchParams();
+  const router = useRouter();
+  const pathname = usePathname();
 
   useEffect(() => {
+    const params = new URLSearchParams(searchParams);
+    if (categoryIds) {
+      params.delete("category");
+      categoryIds.map((category) => {
+        if (category !== undefined)
+          params.append("category", categoryString(category));
+      });
+      router.replace(`${pathname}?${params}`);
+    } else {
+      params.delete("category");
+      router.replace(`${pathname}?${params}`);
+    }
+
+    if (searchValue) {
+      params.delete("description");
+      params.append("description", searchValue);
+      router.replace(`${pathname}?${params}`);
+    } else {
+      params.delete("description");
+      router.replace(`${pathname}?${params}`);
+    }
```

Comment:
This handles two separate effects, it should be separated to two different `useEffect`s.

Reply:
Won't that cause an override of other values? Like if I only enter the searchValue what should I put in the categories?

Reply:
I might not understand either the business logic or the logic written here correctly. What happens here, if I understand it correctly, we re-route the feed by the search params, in a cascading way of categories -> description search. Is this the intended behavior? Am I understanding it incorrectly?

Reply:
Technically could be interchangeable(I don't really care if its categories or description first)

Reply:
What is this effect supposed to do?
`router.replace(`${pathname}?${params}`);` doesn't redirects?
I might need an example of a use-case. I get e.g. `/feed?category="abc"&description="def"`. In this code, we remove the `category` param, add the `category` param with the changed values and then redirect. Then we hit this component again with the same uri.. Will we ever get to the `searchValue` check?
Also, why do all the `delete` and `append` instead of `set`?

Reply:
Parameters changes means uri change means re-render, essentially. From the docs about `replace`: "Perform a client-side navigation to the provided route without adding a new entry into the browser’s history stack."
So a client-side navigation occurs, meaning a re-render. It might not look like a re-render because it's the same component that only gets its props changed, meaning a prop diff is being made.
The main question is, using this method, do you ever get `searchValue` change have any effect? Especially when you also have a `category` param set?
And I'm not sure why would the parsing be any different via `set`, the code here basically does a `set` but in a more complicated way..

Also btw in order to _not_ cause a re-render, there's a [github discussion](https://github.com/vercel/next.js/discussions/18072#discussioncomment-109059) which suggest accessing the `window` directly, which I don't really like unless we absolutely have to. I think we can solve it better here - there's no problem in re-rendering, you should just be aware of it as it will most likely cause part of this code currently to misbehave.

Reply:
I mean yeah when I edit `searchValue` the URL changes and the filter is applied
With `set` it does `category=categoryAcategoryB` according to a string and we parse with `getAll` and `category=categoryA&category=categoryB`

So if we want to avoid accessing `window`, is it better to leave the rerenders like that?

Reply:
Wait, what you want is `categeory=categoryA&category=categoryB`? Or `categeory=categoryA,categoryB`?
I personally prefer the latter, which can be handled with e.g:
```params.set("category", categoryIds);```

And it's kinda weird that `router.replace` doesn't stop the useeffect flow, but it might be because it only takes effect at the end of the render cycle.
Anyway, regarding separating the effect - bear in mind that changes in only one param will cause "changes" (i.e. code execution, even if the final state will look similar) to both parameters.

Full code, in my opinion, should look like:

```
useEffect(() => {
    const params = new URLSearchParams(searchParams);
    params.set("category", categoryIds);
    router.replace(`${pathname}?${params}`);
    onChange({
      categoryIds: categoryIds
    });
  }, [categoryIds, onChange, searchParams, router, pathname]);

useEffect(() => {
    const params = new URLSearchParams(searchParams);
    params.set("description", searchValue);
    router.replace(`${pathname}?${params}`);
    onChange({
      searchValue: searchValue
    });
  }, [searchValue, onChange, searchParams, router, pathname]);
```

If you want, it can also be:

```
useEffect(() => {
    const params = new URLSearchParams(searchParams);
    params.set("category", categoryIds);
    params.set("description", searchValue);
    router.replace(`${pathname}?${params}`);
    onChange({
      categoryIds: categoryIds,
      searchValue: searchValue
    });
  }, [searchValue, categoryIds, onChange, searchParams, router, pathname]);
```

(you can keep the param deletion if you want, I just think it's unnecessary)

Reply:
What we currently have is option A `categeory=categoryA&category=categoryB`
I tried separating similarly to how you did(just with option A in mind) but the `onChange` expects a valid structure with both values so I separated to different effects but still both use that `onChange`

Reply:
Actually after splitting into 2 effects the description isn't being set at all...

Reply:
I understand the problem, it's to do with the `onChange` function. I've checked the branch locally to understand better what exactly is being done here; we basically have two types of states that control the same data, which is not the best approach (and actually hinders "single source of truth"). It'd be better to rely only on the search params as state, and at most mirror it to a single state object (just for the sake of better data control in the code).
I have some changes to reflect what I mean; I can push it here if you want.
By the way, I've noticed that the router change takes a bit to load (especially the first one), and I assume it's because of the massive network load because of all the requests to feed. I never looked into it, are there more requests than needed, or is there just way too much data in the dev environment?

Reply:
If you can push that would be great
The feed makes several requests at a time for each CV(and we also need to remember that we need to fetch the cvs):
1. Blurred preview
2. Updated(and updating) preview
3. CV's uploader
There are caching and memos which I think work as intended but not quite sure...

Reply:
I need to look into the caching mechanisms used, because I don't think they work properly (tho it might be simply dev env-related non-optimization, but I doubt it).
And what do you mean by "updated preview"? and the CV uploader?
And I'll push soon^tm

Reply:
Pushed the commit. Please look through it - it is NOT complete; if you think it should be changed (or not accepted), just write comments on it. I've also added two notes on (what I think are) necessary improvement performance-wise.
Note that this page is the heaviest performance-wise of the site, we need to make sure we keep it as performant as we can.

Reply:
I saw the commit, and it looks good and it's working as intended
Regarding the notes, the debounce would be important, and we can change the formatting if this makes parsing easier.

Do you want me to take this or do you already have a general idea of how things would look like?

Reply:
Planned on letting you handle it, but I just finished my last exam so I have a couple of days to work on the project so I can take it :)

Reply:
Ran the changes locally, and they all worked perfectly, turning debounce to a hook was a nice addition